### PR TITLE
Reused calculated block hash

### PR DESCRIPTION
**Pull request by Vovchyk**
## Description

This change allows to cache a block hash after first call to `BlockHeader.getHash()` and reuse cached value in next calls.

## Motivation and Context

See https://github.com/rsksmart/rskj/issues/1584

## How Has This Been Tested?

unit tests / running node in local env.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

* **Other information**:

### CODE REVIEW

File: rskj-core/src/main/java/org/ethereum/core/BlockHeader.java [hunk 1]
```diff
@@ -293,12 +302,17 @@ public void setLogsBloom(byte[] logsBloom) {
...
             throw new SealedBlockHeaderException("trying to alter logs bloom");
         }
+        this.hash = null;
 
         this.logsBloom = logsBloom;
     }
 
     public Keccak256 getHash() {
-        return new Keccak256(HashUtil.keccak256(getEncoded()));
+        if (this.hash == null) {
```
Comment (donequis) [inline rskj-core/src/main/java/org/ethereum/core/BlockHeader.java:311]: Can I recommend doing it on sealed blocks only?

**Comment by joaquinlpereyra** [comment]
This should work, but potentially opens up the possibility of absolutely destroying a block's hash. I just tried deleting a `this.hash = null` and all tests run correctly. When you delete _all_ of them some tests fail but at first glance have no relationship with the change. 

It'd be interesting to see a test that modifies the block somehow and then checks that the hash has indeed changed.

**Comment by Vovchyk** [comment]
> This should work, but potentially opens up the possibility of absolutely destroying a block's hash.

Yeah, we should be careful here. Btw, similar technique is being used in the `Transaction` class.

> It'd be interesting to see a test that modifies the block somehow and then checks that the hash has indeed changed.

Good point. Added a test for this.

**Review by joaquinlpereyra — APPROVED** [review]
🟢

**Comment by Vovchyk** [comment]
pipeline: run

**Comment by Vovchyk** [comment]
pipeline: run

**Comment by Vovchyk** [comment]
pipeline: run

**Comment by Vovchyk** [comment]
pipeline: run

**Review by heitorfm — APPROVED** [review]

**Comment by Vovchyk** [comment]
Resolves #1585

**Pull request merged**